subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/FileSystemSnapshotterVirtualFileSystem.java

```diff
@@ -0,0 +1,115 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.vfs.impl;
+
+import com.google.common.collect.Interner;
+import org.gradle.internal.file.FileMetadataSnapshot;
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.file.Stat;
+import org.gradle.internal.file.impl.DefaultFileMetadata;
+import org.gradle.internal.hash.FileHasher;
+import org.gradle.internal.hash.HashCode;
+import org.gradle.internal.snapshot.FileSystemLocationSnapshot;
+import org.gradle.internal.snapshot.FileSystemSnapshot;
+import org.gradle.internal.snapshot.FileSystemSnapshotBuilder;
+import org.gradle.internal.snapshot.FileSystemSnapshotVisitor;
+import org.gradle.internal.snapshot.FileSystemSnapshotter;
+import org.gradle.internal.snapshot.MerkleDirectorySnapshotBuilder;
+import org.gradle.internal.snapshot.SnapshottingFilter;
+import org.gradle.internal.snapshot.impl.DefaultFileSystemMirror;
+import org.gradle.internal.snapshot.impl.DefaultFileSystemSnapshotter;
+import org.gradle.internal.vfs.VirtualFileSystem;
+
+import java.io.File;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public class FileSystemSnapshotterVirtualFileSystem implements VirtualFileSystem {
+
+    private final FileSystemSnapshotter snapshotter;
+    private final DefaultFileSystemMirror mirror;
+
+    public FileSystemSnapshotterVirtualFileSystem(FileHasher hasher, Interner<String> stringInterner, Stat stat, String... defaultExcludes) {
+        this.mirror = new DefaultFileSystemMirror();
+        this.snapshotter = new DefaultFileSystemSnapshotter(hasher, stringInterner, stat, mirror, defaultExcludes);
+    }
+
+    @Override
+    public void read(String location, FileSystemSnapshotVisitor visitor) {
+        maybeShortCircuitVisitor(snapshotter.snapshot(new File(location)), visitor);
+    }
+
+    @Override
+    public <T> Optional<T> readRegularFileContentHash(String location, Function<HashCode, T> visitor) {
+        HashCode hashCode = snapshotter.getRegularFileContentHash(new File(location));
+        return Optional.ofNullable(hashCode).map(visitor);
+    }
+
+    @Override
+    public void read(String location, SnapshottingFilter filter, FileSystemSnapshotVisitor visitor) {
+        FileSystemSnapshot snapshot = snapshotter.snapshotDirectoryTree(new File(location), filter);
+        if (snapshot instanceof FileSystemLocationSnapshot) {
+            maybeShortCircuitVisitor((FileSystemLocationSnapshot) snapshot, visitor);
+        } else {
+            snapshot.accept(visitor);
+        }
+    }
+
+    @Override
+    public void update(Iterable<String> locations, Runnable action) {
+        locations.forEach(location -> mirror.beforeOutputChange(locations));
+    }
+
+    @Override
+    public void invalidateAll() {
+        mirror.beforeBuildFinished();
```

Comment:
Hm...

Reply:
I renamed the method on the `mirror` to `invalidateAll()`.